pulsar-package-manager/src/main/java/org/apache/pulsar/packages/manager/impl/PackageImpl.java

```diff
@@ -0,0 +1,146 @@
+import org.apache.pulsar.common.util.FutureUtil;
+import org.apache.pulsar.packages.manager.PackageMetadata;
+import org.apache.pulsar.packages.manager.PackageStorage;
+import org.apache.pulsar.packages.manager.Package;
+import org.apache.pulsar.packages.manager.exception.PackageAlreadyExistsException;
+import org.apache.pulsar.packages.manager.exception.PackageMetaAlreadyExistsException;
+import org.apache.pulsar.packages.manager.exception.PackageMetaNotFoundException;
+import org.apache.pulsar.packages.manager.exception.PackageNotFoundException;
+import org.apache.pulsar.packages.manager.naming.PackageName;
+import org.apache.pulsar.packages.manager.naming.PackageType;
+
+/**
+ * The implementation of package management.
+ */
+public class PackageImpl implements Package {
+
+    private final PackageStorage packageStorage;
+
+    public PackageImpl(PackageStorage packageStorage) {
+        this.packageStorage = packageStorage;
+    }
+
+    @Override
+    public CompletableFuture<PackageMetadata> getMeta(PackageName packageName) {
+        String metadataPath = getMetadataStoragePath(packageName);
+        return packageStorage.existAsync(metadataPath)
+            .thenCompose(exists -> exists
+                ? packageStorage.readAsync(metadataPath)
+                : FutureUtil.failedFuture(new PackageMetaNotFoundException("Package metadata does not existAsync")))
+            .thenApply(meta -> new Gson().fromJson(new String(meta), PackageMetadata.class));
+    }
+
+    @Override
+    public CompletableFuture<Void> updateMeta(PackageName packageName, PackageMetadata metadata) {
+        String metadataPath = getMetadataStoragePath(packageName);
+        return packageStorage.existAsync(metadataPath)
+            .thenCompose(exists -> exists
+                ? packageStorage.writeAsync(metadataPath, new Gson().toJson(metadata).getBytes())
+                : FutureUtil.failedFuture(new PackageMetaNotFoundException("Package metadata does not existAsync")));
+    }
+
+    @Override
+    public CompletableFuture<Void> download(PackageName packageName, OutputStream outputStream) {
+        String pacakgePath = getPackageStoragePath(packageName);
+        return packageStorage.existAsync(pacakgePath)
+            .thenCompose(exists -> exists
+                ? packageStorage.readAsync(pacakgePath, outputStream)
+                : FutureUtil.failedFuture(new PackageNotFoundException("Package does not existAsync")));
+    }
+
+    @Override
+    public CompletableFuture<Void> upload(PackageName packageName, PackageMetadata metadata, InputStream inputStream) {
+        String metadataPath = getMetadataStoragePath(packageName);
+        String packagePath = getPackageStoragePath(packageName);
+        return packageStorage.existAsync(metadataPath)
+            .thenCompose(metaExists -> metaExists
+                ? FutureUtil.failedFuture(new PackageMetaAlreadyExistsException("Package metadata already exists"))
```

Comment:
If a previous upload operation failed in between uploading metadata and upload data, how does a user upload the package again?

Reply:
Currently, If this happening, I think the user needs to delete the package and upload it again.
Do you have some advice?